Faster sequential access for stored fields

jimczi opened the pull request:
Faster sequential access for stored fields

Spinoff of #61806
Today retrieving stored fields at search time is optimized for random access.
So we make no effort to keep state in order to not decompress the same data
multiple times because two documents might be in the same compressed block.
This strategy is acceptable when retrieving a top N sorted by score since
there is no guarantee that documents will be on the same block.
However, we have some use cases where the document to retrieve might be
completely sequential:
* Scrolls or normal search sorted by document id.
* Queries on Runtime fields that extract from _source.

This commit exposes a sequential stored fields reader in the
custom leaf reader that we use at search time.
That allows to leverage the merge instances of stored fields readers that
are optimized for sequential access.
This change focuses on the fetch phase for now and leverages the merge instances
for stored fields only if all documents to retrieve are adjacent.
Applying the same logic in the source lookup of runtime fields should
be trivial but will be done in a follow up.

The speedup on queries sorted by doc id is significant.
I played with the scroll task of the [http_logs rally track](https://elasticsearch-benchmarks.elastic.co/#tracks/http-logs/nightly/default/30d)
on my laptop and had the following result:
```
|                                                        Metric |   Task |    Baseline |   Contender |     Diff |    Unit |
|--------------------------------------------------------------:|-------:|------------:|------------:|---------:|--------:|
|                                            Total Young Gen GC |        |       0.199 |       0.231 |    0.032 |       s |
|                                              Total Old Gen GC |        |           0 |           0 |        0 |       s |
|                                                    Store size |        |     17.9704 |     17.9704 |        0 |      GB |
|                                                 Translog size |        | 2.04891e-06 | 2.04891e-06 |        0 |      GB |
|                                        Heap used for segments |        |    0.820332 |    0.820332 |        0 |      MB |
|                                      Heap used for doc values |        |    0.113979 |    0.113979 |        0 |      MB |
|                                           Heap used for terms |        |     0.37973 |     0.37973 |        0 |      MB |
|                                           Heap used for norms |        |     0.03302 |     0.03302 |        0 |      MB |
|                                          Heap used for points |        |           0 |           0 |        0 |      MB |
|                                   Heap used for stored fields |        |    0.293602 |    0.293602 |        0 |      MB |
|                                                 Segment count |        |         541 |         541 |        0 |         |
|                                                Min Throughput | scroll |     12.7872 |     12.8747 |  0.08758 | pages/s |
|                                             Median Throughput | scroll |     12.9679 |     13.0556 |  0.08776 | pages/s |
|                                                Max Throughput | scroll |     13.4001 |     13.5705 |  0.17046 | pages/s |
|                                       50th percentile latency | scroll |     524.966 |     251.396 |  -273.57 |      ms |
|                                       90th percentile latency | scroll |     577.593 |     271.066 | -306.527 |      ms |
|                                      100th percentile latency | scroll |      664.73 |     272.734 | -391.997 |      ms |
|                                  50th percentile service time | scroll |     522.387 |     248.776 | -273.612 |      ms |
|                                  90th percentile service time | scroll |     573.118 |      267.79 | -305.328 |      ms |
|                                 100th percentile service time | scroll |     660.642 |     268.963 | -391.678 |      ms |
|                                                    error rate | scroll |           0 |           0 |        0 |       % |
```

Closes #62024

elasticmachine commented:
Pinging @elastic/es-search (:Search/Search)

jpountz reviewed (commented):
Woohoo! The change makes sense to me overall although I'm not completely happy with cutting over entirely to CodecReader. I wonder if we should instead introduce a new subclass of LeafReader that introduces a new method such as `getSequentialStoredFieldsReader` and make sure all our security/exitable leaf readers implement it (I haven't fully thought through the implications).

jimczi committed:
Faster sequential access for stored fields

Spinoff of #61806
Today retrieving stored fields at search time is optimized for random access.
So we make no effort to keep state in order to not decompress the same data
multiple times because two documents might be in the same compressed block.
This strategy is acceptable when retrieving a top N sorted by score since
there is no guarantee that documents will be on the same block.
However, we have some use cases where the document to retrieve might be
completely sequential:
* Scrolls or normal search sorted by document id.
* Queries on Runtime fields that extract from _source.

This commit allows to expose all the custom readers that we use at search time
as codec readers in order to be able to leverage the merge instances of
stored fields readers that are optimized for sequential access.
This change focuses on the fetch phase for now and leverages the merge instances
for stored fields only if all documents to retrieve are adjacent.
Applying the same logic in the source lookup of runtime fields should
be trivial but will be done in a follow up.

The speedup on queries sorted by doc id is significant.
I played with the scroll task of the [http_logs rally track](https://elasticsearch-benchmarks.elastic.co/#tracks/http-logs/nightly/default/30d)
on my laptop and had the following result:
```
|                                                        Metric |   Task |    Baseline |   Contender |     Diff |    Unit |
|--------------------------------------------------------------:|-------:|------------:|------------:|---------:|--------:|
|                                            Total Young Gen GC |        |       0.199 |       0.231 |    0.032 |       s |
|                                              Total Old Gen GC |        |           0 |           0 |        0 |       s |
|                                                    Store size |        |     17.9704 |     17.9704 |        0 |      GB |
|                                                 Translog size |        | 2.04891e-06 | 2.04891e-06 |        0 |      GB |
|                                        Heap used for segments |        |    0.820332 |    0.820332 |        0 |      MB |
|                                      Heap used for doc values |        |    0.113979 |    0.113979 |        0 |      MB |
|                                           Heap used for terms |        |     0.37973 |     0.37973 |        0 |      MB |
|                                           Heap used for norms |        |     0.03302 |     0.03302 |        0 |      MB |
|                                          Heap used for points |        |           0 |           0 |        0 |      MB |
|                                   Heap used for stored fields |        |    0.293602 |    0.293602 |        0 |      MB |
|                                                 Segment count |        |         541 |         541 |        0 |         |
|                                                Min Throughput | scroll |     12.7872 |     12.8747 |  0.08758 | pages/s |
|                                             Median Throughput | scroll |     12.9679 |     13.0556 |  0.08776 | pages/s |
|                                                Max Throughput | scroll |     13.4001 |     13.5705 |  0.17046 | pages/s |
|                                       50th percentile latency | scroll |     524.966 |     251.396 |  -273.57 |      ms |
|                                       90th percentile latency | scroll |     577.593 |     271.066 | -306.527 |      ms |
|                                      100th percentile latency | scroll |      664.73 |     272.734 | -391.997 |      ms |
|                                  50th percentile service time | scroll |     522.387 |     248.776 | -273.612 |      ms |
|                                  90th percentile service time | scroll |     573.118 |      267.79 | -305.328 |      ms |
|                                 100th percentile service time | scroll |     660.642 |     268.963 | -391.678 |      ms |
|                                                    error rate | scroll |           0 |           0 |        0 |       % |
```

Closes #62024

jimczi commented on server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:
```diff
@@ -114,7 +118,17 @@ public void execute(SearchContext context) {
+                        // get better sequential access.
+                        storedFieldsReader = storedFieldsReader.getMergeInstance();
+                    } else {
+                        storedFieldsReader = storedFieldsReader.clone();
```
got it, thanks

jimczi commented:
@jpountz , I modified this PR with your idea of having an abstract filter leaf reader that exposes `getSequentialStoredFieldsReader`. The change  is much smaller now, can you take another look ?

jpountz commented on server/src/main/java/org/elasticsearch/common/lucene/index/SequentialStoredFieldsLeafReader.java:
```diff
@@ -0,0 +1,64 @@
+        super(in);
+    }
+
+    protected abstract StoredFieldsReader doGetSequentialStoredFieldsReader(StoredFieldsReader reader);
```
Can you add javadocs?

jimczi replied:
I pushed https://github.com/elastic/elasticsearch/pull/62509/commits/9d3f33fa1ad4877d55bfc1a13acd73111cfb6d46

jpountz commented on server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:
```diff
@@ -524,4 +550,18 @@ private static void fillDocAndMetaFields(SearchContext context, FieldsVisitor fi
+                return false;
+            }
+        }
+        return true;
```
nit: since doc ids are sorted, we should be able to do something like `return docs.length > 0 && docs[docs.length-1].docId - docs[0].docId == docs.length - 1;`

jimczi replied:
++, I pushed https://github.com/elastic/elasticsearch/pull/62509/commits/9d3f33fa1ad4877d55bfc1a13acd73111cfb6d46

jpountz commented on server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -5940,4 +5941,25 @@ public void testNotWarmUpSearcherInEngineCtor() throws Exception {
     }
+
+    public void testProducesStoredFieldsReader() throws Exception {
+        // Make sure that the engine produces a CodecReader.
```
```suggestion
        // Make sure that the engine produces a SequentialStoredFieldsLeafReader.
```

jimczi replied:
https://github.com/elastic/elasticsearch/pull/62509/commits/9d3f33fa1ad4877d55bfc1a13acd73111cfb6d46

jimczi committed:
fix SeqIdGeneratingFilterReader

jpountz commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetReaderTests.java:
```diff
@@ -215,6 +218,41 @@ public void testCoreCacheKey() throws Exception {
         IOUtils.close(ir, ir2, iw, dir);
     }
 
+    public void testProducesCodecReader() throws Exception {
```
```suggestion
    public void testProducesSequentialStoredFieldsLeafReader() throws Exception {
```

jimczi replied:
https://github.com/elastic/elasticsearch/pull/62509/commits/9d3f33fa1ad4877d55bfc1a13acd73111cfb6d46

jpountz reviewed (approved):
I like it, I find it simpler now.

jimczi committed:
feedback

jimczi committed:
unused import
